### PR TITLE
Render NonClusterHost resources

### PR DESCRIPTION
**Pull request by hjiawei**
## Description

This changeset adds a nonclusterhost controller to render NonClusterHost related resources like ServiceAccount, ClusterRole, and ClusterRoleBinding. When a NonClusterHost resource is created in a cluster, SA and token secret will be created automatically inside the `calico-system` namespace. Corresponding ClusterRole and ClusterRoleBindings will also be created for non-cluster host calico-node and calico-fluent-bit agents to talk to the cluster and forward logs to the ingestion endpoint.

## For PR author

- [x] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/nonclusterhost/nonclusterhost_controller.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	return add(mgr, c)
+}
+
+func newReconciler(mgr manager.Manager, opts options.AddOptions) reconcile.Reconciler {
+	r := &ReconcileNonClusterHost{
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+		status: status.New(mgr.GetClient(), "nonclusterhost", opts.KubernetesVersion),
```
Comment (rene-dekker) [inline pkg/controller/nonclusterhost/nonclusterhost_controller.go:63]: We use kebab-case for this: `non-cluster-hosts` (I think plural would probably be my preference).

File: pkg/controller/nonclusterhost/nonclusterhost_controller.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+	instance, err := utils.GetNonClusterHost(ctx, r.client)
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Failed to query NonClusterHost resource", err, logc)
+		return reconcile.Result{}, err
+	}
+
+	ch := utils.NewComponentHandler(logc, r.client, r.scheme, instance)
+
+	if instance == nil {
```
Comment (rene-dekker) [inline pkg/controller/nonclusterhost/nonclusterhost_controller.go:97]: This block is not necessary.

If the resources are created previously, they would have also had an ownerref associated to the NonClusterHost CR. So if that CR is deleted, all associated resources would also be deleted automatically.

File: pkg/render/nonclusterhost/nonclusterhost.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+type nonClusterHostComponent struct {
+	cfg *Config
+}
+
+func (c *nonClusterHostComponent) ResolveImages(is *operatorv1.ImageSet) error {
+	return nil
+}
+
+func (c *nonClusterHostComponent) SupportedOSType() rmeta.OSType {
+	return rmeta.OSTypeLinux
```
Comment (rene-dekker) [inline pkg/render/nonclusterhost/nonclusterhost.go:53]: I think this should be OSTypeAny; no pods are deployed as a result of this.

File: pkg/render/nonclusterhost/nonclusterhost.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+				"stagedglobalnetworkpolicies",
+				"stagedkubernetesnetworkpolicies",
+				"stagednetworkpolicies",
+				"tiers",
+			},
+			Verbs: []string{"get", "list", "watch"},
+		},
+	}
+
+	// calico fluent-bit rules
```
Comment (rene-dekker) [inline pkg/render/nonclusterhost/nonclusterhost.go:162]: Thanks for thoroughly commenting this func!

File: pkg/render/nonclusterhost/nonclusterhost.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+			// The annotation below will result in the auto-creation of spec.data.token.
+			Annotations: map[string]string{
+				"kubernetes.io/service-account.name": NonClusterHostObjectName,
+			},
+		},
+		Type: "kubernetes.io/service-account-token",
+	}
+}
+
+func ClusterRole() *rbacv1.ClusterRole {
```
Comment (hjiawei) [inline pkg/render/nonclusterhost/nonclusterhost.go:101]: Has to be a ClusterRole becuase calico node will read/write some cluster wide resources.

File: pkg/controller/nonclusterhost/nonclusterhost_controller.go (hunk 1 above)
Comment (hjiawei) [reply pkg/controller/nonclusterhost/nonclusterhost_controller.go]: Changed to `non-cluster-hosts`.

File: pkg/controller/nonclusterhost/nonclusterhost_controller.go (hunk 2 above)
Comment (hjiawei) [reply pkg/controller/nonclusterhost/nonclusterhost_controller.go]: This resource deletion part is removed (leaving it to k8s owner reference management). The status is still set to CRNotFound and returned.

File: pkg/render/nonclusterhost/nonclusterhost.go (hunk 3 above)
Comment (hjiawei) [reply pkg/render/nonclusterhost/nonclusterhost.go]: Changed to `OSTypeAny`.

File: pkg/render/node.go
Comment (hjiawei) [inline pkg/render/node.go:1]: Reorder by resource names. No real changes.

**Review by rene-dekker — APPROVED** [review]
lgtm

**Pull request merged**